notes/images/divergence-clock.md

```diff
@@ -0,0 +1,196 @@
+
+1. Identifying concurrent changes.
+2. Order all changes (including concurrent ones)
+
+To make this more concrete, when query reads set of facts we need to identify if some facts are conflicting - updated same state via concurrent changes. This would enable projecting concurrent states that consumer could choose from.
+
+Ordering changes would enable us to produce commit index so that we could perform reconciliation without having to replicate full commit history. Furthermore, we would like to index history such that reconiling query results would incur minimal reads, or more simply we would like to avoid following pointers to random parts of the tree.
+
+### Current Design
+
+Currnt design assumed that every causal triple `{ the, of, is, cause }` MUST have a `cause` that reference another causal triple with the same `{ the, of }` as the original. This constrained implied that all causal references would be colocated in the same subtree as selected facts and therefor require no extra reads to reconcile.
+
+### Problem
+
+Limiting causal references to the same fact lineage has implications on the possible consistency guarantees possible. Specifically it becomes impossible to update multiple facts atomically. Below is an illstrutaion the problem, where two concurrent changes are made one updating `by` and `msg` attributes and another updating only `msg` attribute which ends up in a state in which `by` attribute is from one change and `msg` attribute is from another resulting in missatribution.
+
+Desired behavior would be to reconcile changes such that either only `msg` changes to `"Hi"` or both `msg` and `by` change.
+
+```mermaid
+stateDiagram-v2
+  before: <pre align=left>{<br/>  'id﹕b4be4'﹕{<br/>    by﹕'gozala',<br/>    msg﹕"Hej"<br/>  }<br/>}</pre>
+  after: <pre align=left>{<br/>  'id﹕b4be4'﹕{<br/>    by﹕'cdata',<br/>    msg﹕"Hi"<br/>  }<br/>}</pre>
+
+
+  change: <pre align=left>{<br>  the﹕"msg"<br/>  of﹕ "id﹕b4be4"<br/>  is﹕ "Hi"<br/>}</pre>
+
+  conflict: <pre align=left>{<br>  the﹕"msg"<br/>  of﹕ "id﹕b4be4"<br/>  is﹕ "Hello"<br/>}<br/>{<br>  the﹕"by"<br/>  of﹕ "id﹕b4be4"<br/>  is﹕ "cdata"<br/>}</pre>
+
+  before --> change
+  before --> conflict
+  change --> after
+  conflict --> after
+```
+
+> 💭 I think this may actually be a problem with automerge and other popular CRDT implementations.
+
+## Idea
+
+I think there are two potential ways we could address such limitation that I'll try to describe below
+
+### Partially extending causal reference space
+
+We could lift imposed limitation on causal references e.g instead of requiring that all causal references share same `{ the, of }` we could require that all causal references share attribute namespace (slice of `the` before `/`).
+
+This approach would extend consistency guartees from fact to schema boundary, while still keeping relevant causal information in cousin branches so close by in the tree.
```

Comment:
My intuition is that this is too generalized. The whole reason e.g., PouchDB gives you manual conflict resolution is that you need domain knowledge to say when one version of a structure is in conflict with another.